tests/test_transformer.py

```diff
@@ -1570,6 +1570,68 @@ def foo(x: str) -> None:
     )
 
 
+def test_union_annotation() -> None:
+    node = parse(
+        dedent(
+            """
+            from __future__ import annotations
+
+            class A:
+                ...
+
+            def foo(A: Union[A, None]) -> None:
+                pass
+            """
+        )
+    )
+    TypeguardTransformer(["foo"]).visit(node)
+    assert (
+        unparse(node)
+        == dedent(
+            """
+            from __future__ import annotations
+
+            def foo(A: Union[A, None]) -> None:
+                from typeguard import TypeCheckMemo
+                from typeguard._functions import check_argument_types
+                memo = TypeCheckMemo(globals(), locals())
+                check_argument_types('foo', {'A': (A, Union[A, None])}, memo)
+            """
+        ).strip()
+    )
+
+
+def test_union_annotation_with_or_operator() -> None:
+    node = parse(
+        dedent(
+            """
+            from __future__ import annotations
+
+            class A:
+                ...
+
+            def foo(A: A | None) -> None:
+                pass
+            """
+        )
+    )
+    TypeguardTransformer(["foo"]).visit(node)
+    assert (
+        unparse(node)
+        == dedent(
+            """
+            from __future__ import annotations
+
+            def foo(A: A | None) -> None:
+                from typeguard import TypeCheckMemo
+                from typeguard._functions import check_argument_types
+                memo = TypeCheckMemo(globals(), locals())
+                check_argument_types('foo', {'A': (A, None)}, memo)
```

Comment:
This doesn't look right and it's inconsistent with what we get when using `Union` like in the previous test.

Perhaps I should change the code to return `None` if either of the `BinOp` branches has been erased for now?

Reply:
I'll take a good look at this tomorrow. Some notes:

- The first test passes without your patch
- Did you intentionally name the class and the argument the same? I'm not sure how that makes any sense.
- What, in your opinion, should the result look like?

Reply:
Hi, thanks for the quick review!
> The first test passes without your patch
Yes, I added the first test to compare the output of `Union` and `|` to see if they are consistent. It can be removed but I think that it helps with the conversation.

> Did you intentionally name the class and the argument the same? I'm not sure how that makes any sense.
Yes, I did that intentionally as the easiest way to get a `None` transformation of either branch of the `BinOp`.

> What, in your opinion, should the result look like?
It should look the same as the one for `Union` I think. I do understand that there are some issues with shadowed types though: https://github.com/agronholm/typeguard/issues/337

Reply:
Btw, I haven't tried it yet but maybe one way to get access to the shadowed type inside the function is to do something like:
```
from typing import Type

Type[A].__origin__
```

Reply:
Interesting:
```Python
@typechecked
def foo(int: Union[int, bool]) -> int:
    return int

foo(42)
```
gives
```
TypeError: Union[arg, ...]: each arg must be a type. Got 42.
```

So I guess that the behaviour with `Union` is also incorrect due to the shadowing of the type.

Reply:
How about this: cut out the first test, leave the second in. This PR does demonstrably fix a bug, so let's just get this in. If you'd just add a note in the changelog, I'll merge. Any other union related behaviors should be addressed in a separate issue/PR.

Reply:
I don't think that this PR 'fixes' a bug as it is; it only replaces one bug for another one.

In the example above, only `None` is an accepted value for `foo` whereas probably only value should be accepted (as we should effectively handle `int` as `Any` in this case.

This is demonstrated on this test:

https://github.com/agronholm/typeguard/pull/392/commits/5b5f91f3c5926fd2697b8381813c187e3c526548#diff-00ca1530aa0e4a5f0915df54f28fd180973f39b478af9b8beca7b49884903cd5R675-R685

(currently failing)

Reply:
I think that needs a separate bugfix. This PR could be merged to just fix the transformer issue.

Reply:
@agronholm, I removed the failing test and the classic union test. I'll add an issue about the `|` bug after we merge this.

Reply:
Added https://github.com/agronholm/typeguard/issues/395